cunumeric/array.py

```diff
@@ -1741,6 +1741,63 @@ def astype(
         result._thunk.convert(self._thunk, warn=False)
         return result
 
+    def take(self, indices, axis=None, out=None, mode="raise"):
+        if not np.isscalar(indices):
+            # if indices is a tuple or list, bring sub-tuples to the same shape
+            # and concatenate them
+            indices = convert_to_cunumeric_ndarray(indices)
+
+        if axis is None:
+            self = self.ravel()
+            axis = 0
+        elif axis < 0:
+            axis = self.ndim + axis
+        elif axis >= self.ndim:
+            raise ValueError("axis is out of bounds for array of dimension")
+
+        if mode == "raise":
+            if np.isscalar(indices):
+                if (indices < 0) or (indices >= self.shape[axis]):
+                    raise ValueError("invalid entry in indices array")
+            else:
+                if (indices < 0).any() or (indices >= self.shape[axis]).any():
+                    raise ValueError("invalid entry in indices array")
+        elif mode == "wrap":
+            indices = indices % self.shape[axis]
+        elif mode == "clip":
+            if np.isscalar(indices):
+                if indices >= self.shape[axis]:
+                    indices = self.shape[axis] - 1
+                if indices < 0:
+                    indices = 0
+            else:
+                indices = indices.clip(0, self.shape[axis] - 1)
+        else:
+            raise ValueError(
+                "Invalid mode '{}' for take operation".format(mode)
+            )
+        if self.shape[axis] == 0:
+            if indices.size != 0:
+                raise IndexError(
+                    "Cannot do a non-empty take() from an empty axis."
+                )
+            return self
```

Comment:
I would return a copy of `self` here (although it probably doesn't matter in this case).